From c7ed5ff3fbf7f4e144bfbfbb9efb65b497fabe56 Mon Sep 17 00:00:00 2001
From: uros <uros@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Fri, 27 Jun 2014 08:47:17 +0000
Subject: [PATCH 072/134] 	Backport from mainline 	2014-06-26  Uros
 Bizjak  <ubizjak@gmail.com>

	PR target/61586
	* config/alpha/alpha.c (alpha_handle_trap_shadows): Handle BARRIER RTX.

testsuite/ChangeLog:

	Backport from mainline
	2014-06-26  Uros Bizjak  <ubizjak@gmail.com>

	PR target/61586
	* gcc.target/alpha/pr61586.c: New test.



git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-4_8-branch@212066 138bc75d-0d04-0410-961f-82ee72b054a4
---
 gcc/ChangeLog                            |  8 ++++++++
 gcc/config/alpha/alpha.c                 |  6 ++++++
 gcc/testsuite/ChangeLog                  |  8 ++++++++
 gcc/testsuite/gcc.target/alpha/pr61586.c | 10 ++++++++++
 4 files changed, 32 insertions(+)
 create mode 100644 gcc/testsuite/gcc.target/alpha/pr61586.c

diff --git a/gcc/ChangeLog b/gcc/ChangeLog
index 784d3ca..66248a1 100644
--- a/gcc/ChangeLog
+++ b/gcc/ChangeLog
@@ -1,3 +1,11 @@
+2014-06-27  Uros Bizjak  <ubizjak@gmail.com>
+
+	Backport from mainline
+	2014-06-26  Uros Bizjak  <ubizjak@gmail.com>
+
+	PR target/61586
+	* config/alpha/alpha.c (alpha_handle_trap_shadows): Handle BARRIER RTX.
+
 2014-06-26  Bill Schmidt  <wschmidt@linux.vnet.ibm.com>
 
 	PR target/61542
diff --git a/gcc/config/alpha/alpha.c b/gcc/config/alpha/alpha.c
index 5bd49f9..1fe6eb5 100644
--- a/gcc/config/alpha/alpha.c
+++ b/gcc/config/alpha/alpha.c
@@ -8658,6 +8658,11 @@ alpha_handle_trap_shadows (void)
 			}
 		      break;
 
+		    case BARRIER:
+		      /* __builtin_unreachable can expand to no code at all,
+			 leaving (barrier) RTXes in the instruction stream.  */
+		      goto close_shadow_notrapb;
+
 		    case JUMP_INSN:
 		    case CALL_INSN:
 		    case CODE_LABEL:
@@ -8673,6 +8678,7 @@ alpha_handle_trap_shadows (void)
 		  n = emit_insn_before (gen_trapb (), i);
 		  PUT_MODE (n, TImode);
 		  PUT_MODE (i, TImode);
+		close_shadow_notrapb:
 		  trap_pending = 0;
 		  shadow.used.i = 0;
 		  shadow.used.fp = 0;
diff --git a/gcc/testsuite/ChangeLog b/gcc/testsuite/ChangeLog
index a68b157..1932353 100644
--- a/gcc/testsuite/ChangeLog
+++ b/gcc/testsuite/ChangeLog
@@ -1,3 +1,11 @@
+2014-06-27  Uros Bizjak  <ubizjak@gmail.com>
+
+	Backport from mainline
+	2014-06-26  Uros Bizjak  <ubizjak@gmail.com>
+
+	PR target/61586
+	* gcc.target/alpha/pr61586.c: New test.
+
 2014-06-25  Bill Schmidt  <wschmidt@linux.vnet.ibm.com>
 
 	* gfortran.dg/default_format_denormal_2.f90:  Remove xfail for
diff --git a/gcc/testsuite/gcc.target/alpha/pr61586.c b/gcc/testsuite/gcc.target/alpha/pr61586.c
new file mode 100644
index 0000000..afb1af3
--- /dev/null
+++ b/gcc/testsuite/gcc.target/alpha/pr61586.c
@@ -0,0 +1,10 @@
+/* { dg-do compile } */
+/* { dg-options "-O2 -mieee" } */
+
+void foo (int *dimensions, double **params, int hh)
+{
+  if (params[hh])
+    ;
+  else if (dimensions[hh] > 0)
+    params[hh][0] = 1.0f;
+}
-- 
1.8.5.5

